GraphNeuralNetworks/src/samplers.jl

```diff
@@ -0,0 +1,116 @@
+using GraphNeuralNetworks
+using Graphs
+
+# Define a NeighborLoader structure for sampling neighbors
+struct NeighborLoader
+    graph::GNNGraph             # The input GNNGraph (graph + features from GraphNeuralNetworks.jl)
+    num_neighbors::Vector{Int}  # Number of neighbors to sample per node, for each layer
+    input_nodes::Vector{Int}    # Set of input nodes (starting nodes for sampling)
+    num_layers::Int             # Number of layers for neighborhood expansion
+    batch_size::Union{Int, Nothing}  # Optional batch size, defaults to the length of input_nodes if not given
+    neighbors_cache::Dict{Int, Vector{Int}}  # Cache neighbors to avoid recomputation
+end
+
+# Constructor for NeighborLoader with optional batch size
+function NeighborLoader(graph::GNNGraph; num_neighbors::Vector{Int}, input_nodes::Vector{Int}, num_layers::Int, batch_size::Union{Int, Nothing}=nothing)
+    return NeighborLoader(graph, num_neighbors, input_nodes, num_layers, batch_size === nothing ? length(input_nodes) : batch_size, Dict{Int, Vector{Int}}())
+end
+
+# Function to get cached neighbors or compute them
+function get_neighbors(loader::NeighborLoader, node::Int)
+    if haskey(loader.neighbors_cache, node)
+        return loader.neighbors_cache[node]
+    else
+        neighbors = Graphs.neighbors(loader.graph, node, dir = :in)  # Get neighbors from graph
+        loader.neighbors_cache[node] = neighbors
+        return neighbors
+    end
+end
+
+# Function to sample neighbors for a given node at a specific layer
+function sample_neighbors(loader::NeighborLoader, node::Int, layer::Int)
+    neighbors = get_neighbors(loader, node)
+    if isempty(neighbors)
+        return Int[]
+    else
+        num_samples = min(loader.num_neighbors[layer], length(neighbors))  # Limit to required samples for this layer
+        return rand(neighbors, num_samples)  # Randomly sample neighbors
+    end
+end
+
+function induced_subgraph(graph::GNNGraph, nodes::Vector{Int})
```

Comment:
Since this is a generically useful function, this should go into GNNGraphs.jl, be documented and exported, possibly in a separate PR.

Moreover, since Graphs.jl defines this [interface](https://juliagraphs.org/Graphs.jl/stable/core_functions/operators/#Graphs.induced_subgraph-Union{Tuple{T},%20Tuple{U},%20Tuple{T,%20AbstractVector{U}}}%20where%20{U%3C:Integer,%20T%3C:AbstractGraph}), we should overload `Graphs.induced_interface`.

Notice that we already have the very similar function [sample_neighbors](https://github.com/CarloLucibello/GraphNeuralNetworks.jl/blob/a034753acee5187dc1e1b85e209d6d2a9b2b7e34/GNNGraphs/src/sampling.jl#L2) in GNNGraphs.jl. Let's implement `induced_interface` as well in order to be `Graphs.jl` compliant. At some point we could merge the two functionalities though.

Reply:
I've opened another PR